src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/SessionService.java

```diff
@@ -127,33 +139,117 @@ public Session save(Session session) {
+        userTransactionDataList.add(new UserTransactionData(
+            "session",
+            session.getId(),
+            null,
+            "insert",
+            "delete",
+            0)
+        );
+
+        return userTransactionService.startTransaction(upsertSession.getUserTransactionId(), userTransactionDataList);
+    }
+
+    public UserTransaction updateSession(UpsertSession upsertSession) throws IOException {
+
+        Session session = getSessionById(upsertSession.getId());
+
+        List<HearingPart> hearingParts = hearingPartRepository.findBySessionIn(Arrays.asList(session));
+
+        return areTransactionsInProgress(session, hearingParts)
+            ? userTransactionService.transactionConflicted(upsertSession.getUserTransactionId())
+            : updateWithTransaction(session, upsertSession, hearingParts);
+    }
+
+    private boolean areTransactionsInProgress(Session session, List<HearingPart> hearingParts) {
+        List<UUID> entityIds = hearingParts.stream().map(HearingPart::getId).collect(Collectors.toList());
+        entityIds.add(session.getId());
+
+        return userTransactionService.isAnyBeingTransacted(entityIds.stream().toArray(UUID[]::new));
+    }
+
+    @Transactional
+    public UserTransaction updateWithTransaction(Session session,
+                                                 UpsertSession upsertSession,
+                                                 List<HearingPart> hearingParts) throws IOException {
+        session = updateSessionTime(session, upsertSession);
+        save(session);
+
+        List<UserTransactionData> userTransactionDataList = generateUserTransactionDataList(session, hearingParts);
+
+        UserTransaction ut = userTransactionService.startTransaction(upsertSession.getUserTransactionId(),
+            userTransactionDataList);
+
+        String msg = factsMapper.mapUpdateSessionToRuleJsonMessage(session);
+
+        rulesService.postMessage(ut.getId(), RulesService.UPSERT_SESSION, msg);
+
+        ut = userTransactionService.rulesProcessed(ut);
+
+        return ut;
+    }
+
+    private Session updateSessionTime(Session session, UpsertSession upsertSession) {
+        Optional.ofNullable(upsertSession.getDuration()).ifPresent((d) -> session.setDuration(d));
+        Optional.ofNullable(upsertSession.getStart()).ifPresent((s) -> session.setStart(s));
+
+        return session;
+    }
 
+    private List<UserTransactionData> generateUserTransactionDataList(Session session, List<HearingPart> hearingParts)
+        throws JsonProcessingException {
         List<UserTransactionData> userTransactionDataList = new ArrayList<>();
-        userTransactionDataList.add(new UserTransactionData("session", session.getId(), null, "insert", "delete", 0));
 
-        return userTransactionService.startTransaction(createSession.getUserTransactionId(), userTransactionDataList);
+        userTransactionDataList.addAll(Arrays.asList(new UserTransactionData("session",
```

Comment:
You could change it into variable initialization right now:
'List<UserTransactionData> userTransactionDataList = new ArrayList<>(singletonList('
or similiar

Reply:
>btw I'm starting to think about adding a method that would return us 'id' as a string in one call instead of doing getId().toString() all the time, but that's just a thought not to change it now.

Following the same spirit i'd say that inside the UserTransactionData constructor we could use the method: objectMapper.writeValueAsString(entity) to initialize the 'beforeData' field. But again, that's for future changes! (y)